### PR TITLE
New time-preference feature is included

### PR DESCRIPTION
**Pull request by ankushkushwaha**
As Starflight allow users to set time in which they want to get push -notifications, the changes are made to allow developers to set the time-preferences for push-notifications.

### CODE REVIEW

File: StarFlight/Classes/SCStarFlightPushClient.h [hunk 1]
```diff
@@ -16,7 +16,10 @@ extern NSString *const SCStarFlightClientUUIDNotification;
 - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientSecret;
 
 - (void)registerWithToken:(NSString *)token;
-- (void)registerWithToken:(NSString *)token clientUUID:(nullable NSString *)clientUUID tags:(nullable NSArray<NSString *> *)tags;
+- (void)registerWithToken:(NSString *)token
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.h]: @ankushkushwaha Could you keep the same structure with one line only?

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 2]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
 }
 
 - (void)registerWithToken:(NSString *)token tags:(NSArray<NSString *> *)tags
 {
-    [self registerWithToken:token clientUUID:nil tags:tags];
+    [self registerWithToken:token clientUUID:nil tags:tags timePreferences:nil];
 }
 
-- (void)registerWithToken:(NSString *)token clientUUID:(NSString *)clientUUID tags:(NSArray<NSString *> *)tags
+- (void)registerWithToken:(NSString *)token
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Only one line

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 3]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
+               clientUUID:(NSString *)clientUUID
+                     tags:(NSArray<NSString *> *)tags
+          timePreferences:(NSDictionary *)timePreferencesDict
 {
     if (self.appID && self.clientSecret && token)
     {
         NSString *post = [NSString stringWithFormat:@"action=register&appId=%@&clientSecret=%@&type=ios&token=%@", self.appID, self.clientSecret, token];
-
-        if (clientUUID)
+
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Remove this empty space

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 4]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
         NSString *post = [NSString stringWithFormat:@"action=register&appId=%@&clientSecret=%@&type=ios&token=%@", self.appID, self.clientSecret, token];
-
-        if (clientUUID)
+        
+        if (clientUUID && ![clientUUID  isEqualToString: @""])
         {
             post = [post stringByAppendingString:[NSString stringWithFormat:@"&clientUuid=%@", clientUUID]];
         }
-
+
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Remove this empty space

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 5]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
         }
-
+        
         NSString *tagsToRegister = @"";
         for (NSString *tag in tags)
         {
             tagsToRegister = [tagsToRegister stringByAppendingString:[NSString stringWithFormat:@",%@", tag]];
         }
-
+
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Remove this empty space

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 6]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
         NSString *tagsToRegister = @"";
         for (NSString *tag in tags)
         {
             tagsToRegister = [tagsToRegister stringByAppendingString:[NSString stringWithFormat:@",%@", tag]];
         }
-
+        
         tagsToRegister = [tagsToRegister stringByTrimmingCharactersInSet:[NSCharacterSet characterSetWithCharactersInString:@", "]];
-
+
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Remove this empty space

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 7]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
+            
+            post = [post stringByAppendingString:[NSString stringWithFormat:@"&timePreferences=%@", timePreferencesJsonString]];
+            
+        }
+        
+        // time
         NSData *postData = [post dataUsingEncoding:NSUTF8StringEncoding allowLossyConversion:YES];
         NSString *postLength = [NSString stringWithFormat:@"%lu", (unsigned long)[postData length]];
-
+
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Remove this empty space

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 8]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
+        
         NSMutableURLRequest *request = [[NSMutableURLRequest alloc] init];
         [request setURL:[self starFlightAPIURL]];
         [request setHTTPMethod:@"POST"];
+        [request setValue:[self userAgent]  forHTTPHeaderField:@"User-Agent"];
         [request setValue:postLength forHTTPHeaderField:@"Content-Length"];
         [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Current-Type"];
         [request setHTTPBody:postData];
-
+
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Remove this empty space

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 9]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
         [request setValue:postLength forHTTPHeaderField:@"Content-Length"];
         [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Current-Type"];
         [request setHTTPBody:postData];
-
+        
         self.connection = [[SCHTTPNetwork alloc] init];
         [self.connection startResquest:request success:^(NSData *responseData) {
             NSDictionary *responseObject = [NSJSONSerialization JSONObjectWithData:responseData options: NSJSONReadingMutableContainers error:NULL];
-
+
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Remove this empty space

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 10]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
         [self.connection startResquest:request success:^(NSData *responseData) {
             NSDictionary *responseObject = [NSJSONSerialization JSONObjectWithData:responseData options: NSJSONReadingMutableContainers error:NULL];
-
+            
             if (responseObject)
             {
                 [[NSNotificationCenter defaultCenter] postNotificationName:SCStarFlightClientUUIDNotification object:self userInfo:responseObject];
             }
-
+
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Remove this empty space

File: StarFlight/Classes/SCStarFlightPushClient.m [hunk 11]
```diff
@@ -38,57 +38,71 @@ - (instancetype)initWithAppID:(NSString *)appID clientSecret:(NSString *)clientS
...
-
+        
         if (tagsToRegister && ![tagsToRegister isEqualToString:@""])
         {
             post = [post stringByAppendingString:[NSString stringWithFormat:@"&tags=%@", tagsToRegister]];
         }
-
+        
+        if (timePreferencesDict) {
+            NSData *jsonData = [NSJSONSerialization dataWithJSONObject:timePreferencesDict options:NSJSONWritingPrettyPrinted error:nil];
```
Comment (ramonvasc0) [inline StarFlight/Classes/SCStarFlightPushClient.m]: Check for nil, your code might crash if a nil is found in this method.

**Review by ramonvasc0 — CHANGES_REQUESTED** [review]

**Comment by ankushkushwaha** [comment]
Okay thanks for the review. I will make the requested changes next week.

File: StarFlight/Classes/SCStarFlightPushClient.m (hunk 11 above)
Comment (ankushkushwaha) [reply StarFlight/Classes/SCStarFlightPushClient.m]: Please explain more about this, isn't it checking for nil  `if (timePreferencesDict)`? @ramonvasc0

Comment (ramonvasc0) [reply StarFlight/Classes/SCStarFlightPushClient.m]: @ankushkushwaha Your NSData and NSString initWithData can return nil objects if you have a wrong json for example.

**Comment by ankushkushwaha** [comment]
Hi,

Please review the changes and let me know if anything still remaining @ramonvasc0 

- Ankush

**Review by ramonvasc0 — APPROVED** [review]

**Pull request merged**